### DISCUSSION TITLE
New authenticator and case-sensitive user identifier

### DISCUSSION THREAD

**Discussion by jorluiseptor** [symfony/symfony#44011] [Q&A]
I have a login form which I updated to 5.3 with the new authentication method. The user identifier is the email. I have an issue with user identifier being case-sensitive. If I log in with  foo@email.com as the username, I can log in, but I use Foo@email.com I can't log in. In the db, I save the email as lowercase always. I how can I intercept the login process to convert the user input to lowercase? I can do it with JavaScript at the form, but want to avoid that.

**Comment by alessandro-podo** [comment]
you can lowercase the id in the authenticate function

**Comment by chapterjason** [comment]
Most applications I know stores a second column in the database with the canonicalized email which is basically just the lowercased email.

To canonicalize the email you can create a service for. Then you just use this service to canonicalize the email every time it change and canonicalize the input in the authenticator.

For example Sylius use this approach:

- https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/User/Model/UserInterface.php
- https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/User/Canonicalizer/CanonicalizerInterface.php
